packages/ses/src/error/assert.js

```diff
@@ -56,6 +57,23 @@ const quote = (payload, spaces = undefined) => {
 };
 freeze(quote);
 
+// We may relax this later, but for now only space-separated sequences of
+// sufficiently word-like parts may be embedded into details without quoting.
+const canBeRaw = freeze(/^[\w:-]( ?[\w:-])*$/);
```

Comment:
I don't think this is necessary.  If you're deliberately choosing to output the string without quotation marks, I'm happy to presume you know why you're doing that.  IIUC, the reason `q` outputs the quotation marks is because it's often used in contexts where you might not be entirely sure what's going to get spit out, whereas you'd use `r` in contexts where you (supposedly) know what you're doing.  In other words, the idea is to avoid accidentally generating confusing output, not to protect against an adversarial output source.

Reply:
I think I agree with @FUDCo , but let's investigate before dropping a restriction motivated by security.

I know I can state the threat model addressed by the quoting of `quote`, as well as the threat model addressed by the default redaction vs the quoted non-redaction of `quote`. Can anyone state a threat model that the `canBeRaw` test addresses that would not be addressed if `raw` did not do a `canBeRaw` test?

Reply:
FWIW, leaving aside the `canBeRaw` question, I can also state the threat model difference addressed by the distinction between `quote` and `raw`.

Reply:
I added `canBeRaw` to make sure that e.g. future expansion of `kindOf` output to propagate arbitrary tag strings cannot cause rendering error details to produce misleading text.

Reply:
I don't understand. Can you give an example?

Reply:
Sure. [`checkTagged`](https://github.com/endojs/endo/blob/7a685f59a8f61c63da9dbec5ef78cec0d1b17d3b/packages/patterns/src/patterns/patternMatchers.js#L184) currently rejects tagged records in which the tag is not recognized with ``` X`cannot check unrecognized tag ${q(tag)}: ${tagged}` ```, but we might decide in the future to pass through such values, in which case downstream code that relies on the current rejection of unknown tags and e.g. complains about bad input like ``` X`unacceptable tag: ${raw(tag)} must start with "match:"` ``` might instead encounter a tagged record in which the tag contains line feeds or other control characters, in which case we'd probably like to see
```
unacceptable tag: "foo must start with \"match:\"\nim in ur base, killing ur d00ds\nunacceptable tag: bar" must start with "match:"
```
rather than
```
unacceptable tag: foo must start with "match:"
im in ur base, killing ur d00ds
unacceptable tag: bar must start with "match:"
```

Reply:
This is the attacker-controlled data case, i.e., any string determined by anything outside the current code's reliance set. No such string should ***ever*** be handled by `raw` or `bare`. The lack of quotes leaves the quoting vulnerable to an injection attack regarding human interpretation of the resulting diagnostic string. IOW, all injection attacks are actually quoting confusion attacks. Splicing attacker controlled data into trusted strings should always safely quote the attacker controlled data.

This is actually much of the reasoning that initially led me to bundle quoting with unredaction.

In some systems, they use taint tracking to try to avoid such quoting confusions. Since we don't have that tooling (yet?), a case could still be made that your `canBeRaw` check helps reduce the damage of accidents. That possible argument has me on the fence. If failing the `canBeRaw` check caused an error, we could remove it later. However, since it falls back to quote, for the very good reason you explain, it would be harden to relax the constraint later.

Whether we keep the `canBeRaw` (or `canBeBare`) check or not, we need anyway to document that one should never use `raw`/`bare` on data that is possibly attacker controlled, as the risk of attacker-induced quoting confusion. We should also remember that we would benefit from better tooling anyway, to track attacker controlled data and how it mixes with relied-upon data.

All that said, I still weakly prefer that we remove the `canBeRaw`/`canBeBare` check. But if you strongly prefer to keep it, do so.

That answers all my previous questions. But it did clarify for me that the explanation of `raw` needs to warn about this hazard, so that's my only remaining requested change.